### ISSUE TITLE
Questions about class embed

### ISSUE THREAD

**Issue by haohang96** [bytedance/fc-clip#12]
Hi, Thanks for your awesome work.

I have some questions regarding the class_embed in [code link](https://github.com/bytedance/fc-clip/blob/6c87812a6631f22fc074964593382cd7b207ce6b/fcclip/modeling/transformer_decoder/fcclip_transformer_decoder.py#L495):
`class_embed = self.class_embed(maskpool_embeddings + decoder_output)`

 According to the paper: " **For in-vocabulary classification, following [28], the class embeddings are obtained by mask-pooling the pixel features from the pixel decoder’s final output**".

If following the implementation described in the paper strictly, I think it shall be:
`class_embed = self.class_embed(maskpool_embeddings)`

In addition, I noticed that in the [original implementation](https://github.com/facebookresearch/Mask2Former/blob/9b0651c6c1d5b3af2e6da0589b719c514ec0d69a/mask2former/modeling/transformer_decoder/mask2former_transformer_decoder.py#L436C10-L436C10) of Mask2Former, the class logits is generated by the following codes:
`outputs_class = self.class_embed(decoder_output)`

Is such implementation of fc-clip (self.class_embed(maskpool_embeddings + decoder_output))) to keep the same as the original Mask2Former? If not, could you explain why add the **decoder_output** to **maskpool_embeddings** when generating **class_embed**?

Looking forward to your reply. Thanks

**Comment by cornettoyu** [comment]
Hi,

Thanks a lot for your question. Yes, in our early version implementation, we did exactly as described in the paper, which is `class_embed = self.class_embed(maskpool_embeddings)`

Later we follow ODISE to use the sum of mask pooled embeddings and mask queries when preparing the open-source. But we do not observe significant performance difference among the two design choices.

Best,

**Comment by haohang96** [comment]
Thanks for your reply

I am a bit confused regarding **we follow ODISE to use the sum of mask pooled embeddings and mask queries when preparing the open-source**.

I reviewed the code of ODISE,  it appears that ODISE simply computes the class logits using only the decoder output [# code link](https://github.com/NVlabs/ODISE/blob/2b187e4b2ff4c3d5da342aec2cc234b537720a65/odise/modeling/meta_arch/odise.py#L734C13-L734C13):
`outputs_class = self.class_embed(decoder_output)`

It doesn't seem to use the sum of **mask-pooled embeddings** and **mask queries**. Have I perhaps referred to the wrong section of the ODISE code? 

Thanks

**Comment by cornettoyu** [comment]
I believe I referred to [here](https://github.com/NVlabs/ODISE/blob/e97b06c424c575fec9fc5368dd4b3e050d91abc4/odise/modeling/meta_arch/odise.py#L1000)

But it is still worth mention that we do not see the advantage of summing mask-pooled embeddings and mask queries over mask-pooled embeddings alone

**Comment by haohang96** [comment]
In the code you provided at this [link](https://github.com/NVlabs/ODISE/blob/e97b06c424c575fec9fc5368dd4b3e050d91abc4/odise/modeling/meta_arch/odise.py#L1000), it's evident that ODISE utilizes the sum of mask-pooled embeddings and mask queries.

However, it's important to note that the summed results are employed to refine the 'output_mask' in ODISE, but not the class logits used in FC-CLIP. Am I understanding this correctly?

(P.S. I've realized that summing mask-pooled embeddings and mask queries doesn't have an impact on the results. I'm simply looking to gain a better understanding of the design details behind FC-CLIP. I hope this question isn't causing any inconvenience.)

**Comment by cornettoyu** [comment]
The mask_embed obtained in their code is used for [class prediction](https://github.com/NVlabs/ODISE/blob/2b187e4b2ff4c3d5da342aec2cc234b537720a65/odise/modeling/meta_arch/odise.py#L197) as well. The outout_class you pasted is passed to the PooledMaskEmbed as [pred_logits](https://github.com/NVlabs/ODISE/blob/e97b06c424c575fec9fc5368dd4b3e050d91abc4/odise/modeling/meta_arch/odise.py#L984C72-L984C83) but never used.

Feel free to post if you have any additional question regarding FC-CLIP, but details in ODISE is beyond our support :)

**Comment by haohang96** [comment]
Thanks very much for your patient reply. I understand it.
